new Promise

```
const promise = new Promise((resolve, reject) => { // 执行器
  // 异步操作代码...
})
```


Promise 作为一个构造函数，通过 new 操作符可创建一个 promise 实例

该构造函数接收一个函数作为参数，这个函数我们通常称为执行器（executor）函数

执行器函数接收两个参数：这两个参数作为函数可调用并可以传入相应数据，通常我们以resolve和reject进行命名


创建 Promise 类

根据上述示例与描述，现在可以先构造一个初步的 Promise 类

- Promise 具备三种状态：pending、fulfilled（有时候也称为 resolved ）、rejected。
- 构造器内定义 executor，接收实例传过来的执行器函数
- 初始状态为 pending。
- 初始 value （终值）， 它是任何一个合法的 JavaScript 值。
- 初始 reason (拒因) 


```
/ 定义 Promise 的三种状态
const PENDING = 'pending'
const FULFILLED = 'fulfilled'
const REJECTED = 'rejected'

// 定义 MyPromise 类
class MyPromise { 
  constructor(executor) {
    // 初始状态为 pending
    this.status = PENDING
    
    // 初始 value 与 reason
    this.value = undefined
    this.reason = undefined
  }
}

```

resolve 与 reject

回顾一下调用这两个函数的作用

```
const promise = new Promise((resolve, reject) => {
    // 异步操作代码
  setTimeout(() => {
    const rand = Math.random() * 100
    if (rand >= 50) {
        // 成功时执行
      resolve('success')
    } else {
        // 失败时执行
      reject('error')
    }
  }, 1000)
})


```

resolve函数：将 promise 实例的状态从 pending 变为 fulfilled，在异步操作成功时调用，并将异步操作的结果，作为参数进行传递

reject函数： 将 promise 实例的状态从 pending 变为 rejected，在异步操作失败时调用，并将异步操作报出的错误，作为参数进行传递

MyPromise 类中，我们定义 resolve 函数 和 reject 函数，并在函数体内进行 status、value 或 reason的更新，同时调用 executor 执行器，传入 resolve 和 reject

```
// ...
class MyPromise {
  constructor(executor) {
    this.status = PENDING
    this.value = undefined
    this.reason = undefined
    
    // 定义 resolve
    const resolve = value => {
      // 更新状态
      this.status = FULFILLED
      // 更新 value
      this.value = value
    }
    // 定义 reject
    const reject = reason => {
      // 更新状态
      this.status = REJECTED
      // 更新 reason
      this.reason = reason
    }
    // 调用执行器，将 resolve 与 reject 作为参数传递出去
    executor(resolve, reject)
  }
}
    

```

测试

```javascript
// 通过 MyPromise 生成 p1 实例， 执行 resolve 并传递 success 参数
const p1 = new MyPromise((resolve, reject) => {
  resolve('success')
})
console.log(p1) // MyPromise {status: 'fulfilled', value: 'success', reason: undefined}

// 通过 MyPromise 生成 p2 实例， 执行 reject 并传递 error 参数
const p2 = new MyPromise((resolve, reject) => {
  reject('error')
})
console.log(p2) // MyPromise {status: 'rejected', value: undefined, reason: 'error'}

```
